src/db.rs

```diff
@@ -2,154 +2,218 @@
+    /// per each "publicIP:clientID" tuple.
+    /// For example:
+    ///
+    /// "88.22.170.96": [
+    ///     "e7ce02eaa73da35bddea00c82124c7fbbe49b731",
+    ///     "2b3e83cca3ee12c8b41d86bfeca6034ea8cb9056"
+    /// ]
+    ///
+    /// "88.22.170.96:e7ce02eaa73da35bddea00c82124c7fbbe49b731": "message1"
+    /// "88.22.170.96:2b3e83cca3ee12c8b41d86bfeca6034ea8cb9056": "message2"
+    ///
+    /// Each "publicIP:clientID" tuple has a ttl of 115 seconds.
+    ///
+    pub fn set(&self, record: Record) -> RedisResult<()> {
+        let key = format!("{}:{}", record.public_ip, record.client);
+
+        // We need to start watching the keys we care about (public_ip and
+        // public_ip:client) so that our exec fails if the key changes.
+        let _: () = try!(
+            cmd("WATCH").arg(key.clone())
+                        .arg(record.public_ip.clone())
+                        .query(&self.connection)
+        );
+
+        // We check if there's already an entry for this public IP.
+        let is_member: isize = try!(
+            cmd("SISMEMBER").arg(record.public_ip.clone())
+                            .arg(record.client.clone())
+                            .query(&self.connection)
+        );
+
+        if is_member == 0 {
+            // If there is no previous entry for this public IP, we add one
+            // and add the message corresponding to this key (IP:user tuple)
+            info!("{} is not a member of {} yet",
+                  record.client.clone(), record.public_ip.clone());
+            let _: () = try!(
+                pipe().atomic()
+                      .cmd("SADD").arg(record.public_ip.clone())
+                                  .arg(record.client.clone())
+                                  .ignore()
+                      .cmd("SET").arg(key.clone())
+                                 .arg(record.message.clone())
+                                 .query(&self.connection)
+            );
+        } else {
+            // Otherwise, we just update the message from the existing
+            // entry.
+            info!("{} is already a member of {}",
+                  record.client.clone(), record.public_ip.clone());
+            let _: () = try!(
+                cmd("SET").arg(key.clone())
+                          .arg(record.message.clone())
+                          .query(&self.connection)
+            );
+        }
 
-    #[cfg(test)]
-    pub fn clear(&self) -> rusqlite::Result<()> {
-        self.connection.execute_batch(
-            "DELETE FROM boxes;
-             VACUUM;"
-        )
+        // And set the TTL of the message.
+        // The entry in the list of clients for this public IP will be
+        // cleaned up during the .get call. It does no harm to keep it
+        // around until that point.
+        let _: () = try!(
+            cmd("EXPIRE").arg(key.clone())
+                         .arg(115) // 115 seconds.
```

Comment:
Ah right, we were refreshing the timestamp, so we were fine in that case. Ignore my previous comment.

What is not entirely right in the current solution is that we have to wait 10 minutes to see the box go away from the reg server, which makes development a bit harder. So I think a shorter eviction time is better.